### PR TITLE
PKCE improvements

### PR DESCRIPTION
**Pull request by StanWarz**
[[SDK core] Enforce PKCE code flow in checkSession](https://app.asana.com/0/982150578058310/1189474797301249/f)
[[SDK Core] Use PKCE by default in all code flows](https://app.asana.com/0/982150578058310/1184044175244937/f)
Quand cette PR est validée, je releaserai le sdk-core pour corriger la PR sur la sandbox : https://github.com/ReachFive/reachfive-sandbox/pull/63

### CODE REVIEW

File: src/main/apiClient.ts [hunk 1]
```diff
@@ -93,7 +93,7 @@ export type ApiClientConfig = {
   language?: string
   scope?: string
   sso: boolean
-  pkceEnabled?: boolean
+  pkceEnforced?: boolean
```
Comment (axuan25) [inline src/main/apiClient.ts:96]: The value should always be present in theory. The `?` was probably used when the feature was first integrated.

File: src/main/apiClient.ts [hunk 2]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
   }
 
   loginFromSession(opts: AuthOptions = {}): Promise<void> {
-    if (!this.config.sso && !opts.idTokenHint) {
+    if (!this.config.sso && !opts.idTokenHint)
```
Comment (axuan25) [inline src/main/apiClient.ts:186]:
```suggestion
    if (this.config.sso)
```
The previous condition is no longer true. We cannot call loginFromSession and/or checkSession using ID Tokens.

File: src/main/apiClient.ts [hunk 3]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
...
-        useWebMessage: false
+        useWebMessage: false,
       }),
       prompt: 'none'
     })
   }
 
   checkSession(opts: AuthOptions = {}): Promise<AuthResult> {
-    if (!this.config.sso && !opts.idTokenHint) {
+    if (!this.config.sso && !opts.idTokenHint)
```
Comment (axuan25) [inline src/main/apiClient.ts:201]:
```suggestion
    if (this.config.sso)
```
Same as above. We need to remove this asap as it's been fixed at the backend a while ago.

File: src/main/apiClient.ts [hunk 4]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
...
       return Promise.reject(
         new Error("Cannot call 'checkSession' without 'idTokenHint' parameter if SSO is not enabled.")
```
Comment (axuan25) [inline src/main/apiClient.ts:203]:
```suggestion
        new Error("Cannot call 'checkSession' if SSO is not enabled.")
```

File: src/main/apiClient.ts [hunk 5]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
   }
 
   loginFromSession(opts: AuthOptions = {}): Promise<void> {
-    if (!this.config.sso && !opts.idTokenHint) {
+    if (!this.config.sso && !opts.idTokenHint)
       return Promise.reject(
         new Error("Cannot call 'loginFromSession' without 'idTokenHint' parameter if SSO is not enabled.")
```
Comment (axuan25) [inline src/main/apiClient.ts:188]:
```suggestion
        new Error("Cannot call 'loginFromSession' if SSO is not enabled.")
```

File: src/main/apiClient.ts [hunk 6]
```diff
@@ -809,14 +816,10 @@ export default class ApiClient {
...
-      if (authParams.responseType === 'token')
-        return Promise.reject(new Error('Cannot use implicit flow when PKCE is enabled'))
-      else
-        return computePkceParams()
-    } else
-      return Promise.resolve({})
+  private static buildPkceParams(authParams: AuthParameters): Promise<PkceParams | {}> {
+    if (authParams.responseType === 'token') {
+      return Promise.reject(new Error('Cannot use implicit flow when PKCE is enforced'))
+    } else return computePkceParams()
```
Comment (axuan25) [inline src/main/apiClient.ts:822]: 1. Nitpick on terminology: I prefer sticking to "get" as we aren't building the PKCE params using any provided arguments
2. The error states implicit cannot be used when PKCE is enforced but does not check against the configuration

In this method, we need to check that:
```javascript
if (responseType === 'code') {
  computePkceParams()
} else {
  if (authParams.responseType === 'token' && this.config.pkceEnforced) {
    return Promise.reject(new Error('Cannot use implicit flow when PKCE is enforced'))
  } else {
    return Promise.resolve({})
  }
}
```
_(after which the returned promise will still be `Promise<PkceParams | {}>` whereas it is currently just `Promise<PkceParams>`)_

File: src/main/apiClient.ts [hunk 7]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
...
+    })
   }
 
   private getWebMessage(
-    src: string,
-    origin: string,
-    redirectUri?: string,
+      src: string,
+      origin: string,
+      redirectUri: string,
```
Comment (axuan25) [inline src/main/apiClient.ts:232]:
```suggestion
    src: string,
    origin: string,
    redirectUri: string,
```

**Review by axuan25 — CHANGES_REQUESTED** [review]
We can perhaps add some tests to ensure the new behavior

File: CHANGELOG.md [hunk 8]
```diff
@@ -117,7 +117,7 @@ The [`logout`](https://developer.reach5.co/api/identity-web/#logout) method was
 ### 02/12/2019
 
 ### Fixes
-Add the missing implementation of `Buffer` to address an issue with the UMD bundle when PKCE is enabled.
+Add the missing implementation of `Buffer` to address an issue with the UMD bundle when PKCE is enforced.
```
Comment (axuan25) [inline CHANGELOG.md:120]: Back in early 2019, PKCE actually was either enabled or disabled. The change to "enforce" it came much later.

File: src/main/apiClient.ts [hunk 9]
```diff
@@ -810,13 +817,13 @@ export default class ApiClient {
...
-        return computePkceParams()
-    } else
-      return Promise.resolve({})
+    if (authParams.responseType === 'code')
+      computePkceParams()
+
+    if (authParams.responseType === 'token' && this.config.pkceEnforced)
+      return Promise.reject(new Error('Cannot use implicit flow when PKCE is enforced'))
+
+    return Promise.resolve({})
```
Comment (axuan25) [inline src/main/apiClient.ts:826]: Could we stick to complete expressions (if-else) instead of early returns?

File: src/main/apiClient.ts [hunk 10]
```diff
@@ -810,13 +817,13 @@ export default class ApiClient {
...
   private getPkceParams(authParams: AuthParameters): Promise<PkceParams | {}> {
-    if (this.config.pkceEnabled) {
-      if (authParams.responseType === 'token')
-        return Promise.reject(new Error('Cannot use implicit flow when PKCE is enabled'))
-      else
-        return computePkceParams()
-    } else
-      return Promise.resolve({})
+    if (authParams.responseType === 'code')
+      computePkceParams()
```
Comment (axuan25) [inline src/main/apiClient.ts:821]:
```suggestion
      return computePkceParams()
```

File: src/main/apiClient.ts [hunk 11]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
...
+        opts.redirectUri || "",
+      )
+    })
   }
 
   private getWebMessage(
     src: string,
     origin: string,
-    redirectUri?: string,
+    redirectUri: string,
```
Comment (axuan25) [inline src/main/apiClient.ts:232]:
```suggestion
```

File: src/main/apiClient.ts [hunk 12]
```diff
@@ -250,7 +257,7 @@ export default class ApiClient {
           if (result.code) {
             resolve(this.exchangeAuthorizationCodeWithPkce({
               code: result.code,
-              redirectUri: redirectUri || ""
+              redirectUri,
```
Comment (axuan25) [inline src/main/apiClient.ts:260]:
```suggestion
              redirectUri: location.origin,
```

File: src/main/apiClient.ts [hunk 13]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
...
+        ...authParams,
+        ...maybeChallenge,
+      }
+
+      const authorizationUrl = this.getAuthorizationUrl(params)
+
+      return this.getWebMessage(
+        authorizationUrl,
+        `https://${this.config.domain}`,
+        opts.redirectUri || "",
```
Comment (axuan25) [inline src/main/apiClient.ts:224]:
```suggestion
```

File: src/main/apiClient.ts [hunk 14]
```diff
@@ -463,17 +469,18 @@ export default class ApiClient {
...
-        ...pick(tkn, 'tkn')
+        ...pick(tkn, 'tkn'),
       })
 
       if (auth.useWebMessage) {
         return this.getWebMessage(
           `${this.authorizeUrl}?${queryString}`,
           `https://${this.config.domain}`,
-          auth.redirectUri || ""
+          auth.redirectUri || "",
```
Comment (axuan25) [inline src/main/apiClient.ts:483]:
```suggestion
```

**Review by axuan25 — COMMENTED** [review]
Last change suggestion:

Since passing a `redirect_uri` doesn't make much sense for our SDK users, we can remove the parameter altogether and hardcode `redirectUri` to `location.origin`.

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## Not released yet
+
+### Features
+Now use PKCE in every convenient API call which does not lead to a redirection.
+
+### Fixes
+Renamed pkceEnabled into pkceEnforced for the sake of comprehension.
```
Comment (axuan25) [inline CHANGELOG.md:9]:
```suggestion
Generate a PKCE code_challenge by default in all authorization code flows (`responseType === 'code'`)
Use the authorization code flow in checkSession

### Fixes
Renamed erroneously named config value `pkceEnabled` to `pkceEnforced`.
```

**Review by axuan25 — APPROVED** [review]
LGTM

File: src/main/__tests__/remoteSettingsResolving.spec.ts [hunk 16]
```diff
@@ -20,7 +20,8 @@ test('remote settings language has priority over transmitted language', async ()
     {
       clientId,
       domain,
-      language: submittedLanguage
+      language: submittedLanguage,
+      pkceEnforced: false, // TODO STAN
```
Comment (axuan25) [inline src/main/__tests__/remoteSettingsResolving.spec.ts:24]: 👁️

**Review by axuan25 — APPROVED** [review]

File: src/main/__tests__/remoteSettingsResolving.spec.ts [hunk 17]
```diff
@@ -40,7 +41,7 @@ test('remote settings language has priority over transmitted language', async ()
   })
 
   // Then
-  expect(passwordLoginCall).toHaveBeenCalledWith(`https://${domain}/identity/v1/me?fields=name`, {
+  expect(passwordLoginCall).toHaveBeenCalledWith(`https://${domain}/identity/v1/userinfo?fields=name`, {
```
Comment (roxanemace) [inline src/main/__tests__/remoteSettingsResolving.spec.ts:44]: 👍

File: src/main/__tests__/remoteSettingsResolving.spec.ts (hunk 16 above)
Comment (roxanemace) [inline src/main/__tests__/remoteSettingsResolving.spec.ts:24]: ?

**Review by roxanemace — APPROVED** [review]

File: src/main/__tests__/remoteSettingsResolving.spec.ts (hunk 17 above)
Comment (StanWarz) [reply src/main/__tests__/remoteSettingsResolving.spec.ts:44]: 🙏

File: src/main/apiClient.ts [hunk 18]
```diff
@@ -183,46 +183,53 @@ export default class ApiClient {
...
+        opts.redirectUri,
+      )
+    })
   }
 
   private getWebMessage(
     src: string,
     origin: string,
-    redirectUri?: string,
+    redirectUri ?: string,
```
Comment (axuan25) [inline src/main/apiClient.ts:232]:
```suggestion
    redirectUri?: string,
```

**Pull request merged**